plugin/src/main/java/org/opensearch/ml/model/MLModelAndNodeManager.java

```diff
@@ -0,0 +1,294 @@
+        MLIndicesHandler mlIndicesHandler,
+        MLTaskManager mlTaskManager,
+        MLModelCacheHelper modelCacheHelper,
+        MLEngine mlEngine
+    ) {
+        this.clusterService = clusterService;
+        this.client = client;
+        this.threadPool = threadPool;
+        this.xContentRegistry = xContentRegistry;
+        this.modelHelper = modelHelper;
+        this.nodeHelper = nodeHelper;
+        this.modelCacheHelper = modelCacheHelper;
+        this.mlStats = mlStats;
+        this.mlCircuitBreakerService = mlCircuitBreakerService;
+        this.mlIndicesHandler = mlIndicesHandler;
+        this.mlTaskManager = mlTaskManager;
+        this.mlEngine = mlEngine;
+
+        enableAutoReLoadModel = ML_COMMONS_MODEL_AUTO_RELOAD_ENABLE.get(settings);
+        clusterService
+            .getClusterSettings()
+            .addSettingsUpdateConsumer(ML_COMMONS_MODEL_AUTO_RELOAD_ENABLE, it -> enableAutoReLoadModel = it);
+    }
+
+    public void autoReLoadModel() {
+        log.debug("enableAutoReLoadModel {} ", enableAutoReLoadModel);
+
+        // if we don't need to reload automatically, just return without doing anything
+        if (!enableAutoReLoadModel) {
+            return;
+        }
+
+        // At opensearch startup, get local node id, if not ml node,we ignored, just return without doing anything
+        String localNodeId = clusterService.localNode().getId();
+        DiscoveryNode node = nodeHelper.getNode(localNodeId);
+        if (!MLNodeUtils.isMLNode(node)) {
+            return;
+        }
+
+        // According to the node id to query the number of retries, if more than 2 (the maximum number of retries), we do not need to retry,
+        // that the number of unsuccessful reload has reached the maximum number of times, do not need to reload
+        Integer reTryTimes = 0;
+        if (isExistedIndex(ML_MODEL_RELOAD_INDEX)) {
+            reTryTimes = getReTryTimes(localNodeId);
+        }
+        if (reTryTimes > ML_MODEL_RELOAD_MAX_RETRY_TIMES) {
+            throw new RuntimeException("have exceeded max retry times, always failure");
+        }
+
+        // auto reload all models of this local node, if it fails, reTryTimes+1, if it succeeds, reTryTimes is cleared to 0
+        autoReLoadModelByNodeId(localNodeId);
+    }
+
+    @VisibleForTesting
+    void autoReLoadModelByNodeId(String nodeId) {
+        SearchRequest searchRequest = new SearchRequest(ML_TASK_INDEX);
+        SearchResponse response = client.execute(SearchAction.INSTANCE, searchRequest).actionGet();
+
+        if (response == null) {
+            return;
+        }
+        SearchHits searchHits = response.getHits();
+        if (searchHits == null) {
+            return;
+        }
+        SearchHit[] hits = searchHits.getHits();
+        if (CollectionUtils.isEmpty(hits)) {
+            return;
+        }
+
+        Integer reTryTimes = 0;
+        for (SearchHit hit : hits) {
```

Comment:
Instead try each record in the result, it's better to aggregate the result first to figure out which models should be reloaded and then arrange several parallel tasks to accomplish this.

Reply:
see the comment above.